One of my engineers using the code from this PR is reporting the dreaded `plexus-utils-1.1.jar` not found error:
```

[ERROR] Failed to execute goal org.teiid:vdb-maven-plugin:1.3-SNAPSHOT:vdb (default-vdb) on project metrics-vdb: Execution default-vdb of goal org.teiid:vdb-maven-plugin:1.3-SNAPSHOT:vdb failed: A required class was missing while executing org.teiid:vdb-maven-pl
ugin:1.3-SNAPSHOT:vdb: org/apache/commons/compress/archivers/ArchiveEntry
[ERROR] -----------------------------------------------------
[ERROR] realm =    extension>org.teiid:vdb-maven-plugin:1.3-SNAPSHOT
[ERROR] strategy = org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy
[ERROR] urls[0] = file:/C:/Users/Ted/.m2/repository/org/teiid/vdb-maven-plugin/1.3-SNAPSHOT/vdb-maven-plugin-1.3-SNAPSHOT.jar
[ERROR] urls[1] = file:/C:/Users/Ted/.m2/repository/org/codehaus/plexus/plexus-utils/1.1/plexus-utils-1.1.jar
[ERROR] Number of foreign imports: 1
[ERROR] import: Entry[import  from realm ClassRealm[maven.api, parent: null]]
[ERROR]
[ERROR] -----------------------------------------------------: org.apache.commons.compress.archivers.ArchiveEntry
```

I'm not sure what I did to suddenly have this crop up...

--- a/src/main/java/org/teiid/VdbMojo.java
+++ b/src/main/java/org/teiid/VdbMojo.java
@@ -21,7 +21,13 @@ import java.io.FileNotFoundException;
 import java.io.FileOutputStream;
 import java.io.FilenameFilter;
 import java.io.IOException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Objects;
 import java.util.Set;
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipInputStream;
@@ -38,6 +44,7 @@ import org.apache.maven.plugin.MojoExecutionException;
 import org.apache.maven.plugin.MojoFailureException;
 import org.apache.maven.plugins.annotations.Mojo;
 import org.apache.maven.plugins.annotations.Parameter;
+import org.apache.maven.plugins.annotations.ResolutionScope;
 import org.apache.maven.project.MavenProject;
 import org.teiid.adminapi.DataPolicy;
 import org.teiid.adminapi.impl.DataPolicyMetadata;
@@ -49,124 +56,132 @@ import org.teiid.adminapi.impl.VDBMetadataParser;
  * https://stackoverflow.com/questions/1427722/how-do-i-create-a-new-packaging-type-for-maven
  * http://softwaredistilled.blogspot.com/2015/07/how-to-create-custom-maven-packaging.html
  */
-@Mojo(name = "vdb")
+@Mojo(name = "vdb", requiresDependencyResolution = ResolutionScope.COMPILE)
 public class VdbMojo extends AbstractMojo {
     private static final String SLASH = "/";
 
-    @Parameter( defaultValue = "${basedir}/src/main/vdb/META-INF/vdb.xml" )
-    private String vdbXmlFile;
+    @Parameter(defaultValue = "${basedir}/src/main/vdb/META-INF/vdb.xml")
+    private File vdbXmlFile;
 
-    @Parameter( defaultValue = "${basedir}/src/main/vdb" )
-    private String vdbFolder;
+    @Parameter(defaultValue = "${basedir}/src/main/vdb")
+    private File vdbFolder;
 
-    @Parameter( defaultValue = "${project}", readonly = true )
+    @Parameter(defaultValue = "${project}", readonly = true)
     private MavenProject project;
 
     @Parameter(property = "project.build.directory", readonly = true)
-    private String outputDirectory;
+    private File outputDirectory;
 
     @Parameter(property = "project.build.finalName", readonly = true)
     private String finalName;
 
-    
-     //A list of folders or files to be included in the final artifact archive.
+    //A list of folders or files to be included in the final artifact archive.
     @Parameter
     private File[] includes;
 
     @Override
     public void execute() throws MojoExecutionException, MojoFailureException {
 
-        File artifact = new File(this.outputDirectory, this.finalName+".vdb");
+        ClassLoader oldCL = Thread.currentThread().getContextClassLoader();
+        ClassLoader pluginClassloader = getClassLoader();
 
-        this.project.getArtifact().setFile(artifact);
+        try {
+            Thread.currentThread().setContextClassLoader(pluginClassloader);
+            File artifact = new File(this.outputDirectory, this.finalName + ".vdb");
 
-        try (ArchiveOutputStream archive = this.getStream(artifact)) {
-            File vdb = this.getVDBFile();
-            if (vdb == null) {
-                throw new MojoExecutionException("No VDB File found in directory" + this.vdbFolder);
-            }
+            this.project.getArtifact().setFile(artifact);
 
-            // add config, classes, lib and META-INF directories
-            File f = new File(this.vdbFolder);
-            Set<File> directories = new LinkedHashSet<>();
-            gatherContents(f, directories);
+            try (ArchiveOutputStream archive = this.getStream(artifact)) {
+                File vdb = this.getVDBFile();
+                if (vdb == null) {
+                    throw new MojoExecutionException("No VDB File found in directory" + this.vdbFolder);
+                }
 
-            // do not allow vdb-import in the case of VDB represented with .ddl
-            if (vdb.getName().endsWith("-vdb.ddl")) {
-            	addFile(archive, "META-INF/vdb.ddl", vdb);
-            	return;
-            }
+                // add config, classes, lib and META-INF directories
+                Set<File> directories = new LinkedHashSet<>();
+                gatherContents(this.vdbFolder, directories);
 
-            // check if the VDB has any vdb imports, if yes, then check the dependencies
-            VDBMetaData top = VDBMetadataParser.unmarshell(new FileInputStream(vdb));
-            if (!top.getVDBImports().isEmpty()) {
-                // read import vdbs
-                Set<Artifact> dependencies = project.getDependencyArtifacts();
-                for (Artifact d : dependencies) {
+                // do not allow vdb-import in the case of VDB represented with .ddl
+                if (vdb.getName().endsWith("-vdb.ddl")) {
+                    addFile(archive, "META-INF/vdb.ddl", vdb);
+                    return;
+                }
 
-                    if (d.getFile() == null || !d.getFile().getName().endsWith(".vdb")) {
-                        continue;
-                    }
+                // check if the VDB has any vdb imports, if yes, then check the dependencies
+                VDBMetaData top = VDBMetadataParser.unmarshell(new FileInputStream(vdb));
+                if (!top.getVDBImports().isEmpty()) {
+                    // read import vdbs
+                    Set<Artifact> dependencies = project.getDependencyArtifacts();
+                    for (Artifact d : dependencies) {
 
-                    File vdbDir = unzipContents(d);
-                    File childFile = new File(vdbDir, "META-INF/vdb.xml");
-                    System.out.println("Merging VDB " + childFile.getCanonicalPath());
-                    VDBMetaData child = VDBMetadataParser.unmarshell(new FileInputStream(childFile));
+                        if (d.getFile() == null || !d.getFile().getName().endsWith(".vdb")) {
+                            continue;
+                        }
 
-                    if (!child.getVDBImports().isEmpty()) {
-                        throw new MojoExecutionException("Nested VDB imports are not supported" + d.getArtifactId());
-                    }
+                        File vdbDir = unzipContents(d);
+                        File childFile = new File(vdbDir, "META-INF/vdb.xml");
+                        getLog().info("Merging VDB " + childFile.getCanonicalPath());
+                        VDBMetaData child = VDBMetadataParser.unmarshell(new FileInputStream(childFile));
 
-                    VDBImportMetadata matched = null;
-                    for (VDBImportMetadata importee : top.getVDBImports()) {
-                        if (child.getName().equals(importee.getName())
-                                && child.getVersion().equals(importee.getVersion())) {
+                        if (!child.getVDBImports().isEmpty()) {
+                            throw new MojoExecutionException("Nested VDB imports are not supported" + d.getArtifactId());
+                        }
 
-                            gatherContents(vdbDir, directories);
+                        VDBImportMetadata matched = null;
+                        for (VDBImportMetadata importee : top.getVDBImports()) {
+                            if (child.getName().equals(importee.getName())
+                                    && child.getVersion().equals(importee.getVersion())) {
 
-                            top.getVisibilityOverrides().putAll(child.getVisibilityOverrides());
-							child.getModelMetaDatas().forEach((k, v) -> {
-								top.addModel(v);
-								String visibilityOverride = top.getPropertyValue(v.getName() + ".visible"); //$NON-NLS-1$
-								if (visibilityOverride != null) {
-									boolean visible = Boolean.valueOf(visibilityOverride);
-									top.setVisibilityOverride(v.getName(), visible);
-								}
-							});
+                                gatherContents(vdbDir, directories);
 
-                            child.getOverrideTranslatorsMap().forEach((k,v) -> top.addOverideTranslator(v));
+                                top.getVisibilityOverrides().putAll(child.getVisibilityOverrides());
+                                child.getModelMetaDatas().forEach((k, v) -> {
+                                    top.addModel(v);
+                                    String visibilityOverride = top.getPropertyValue(v.getName() + ".visible"); //$NON-NLS-1$
+                                    if (visibilityOverride != null) {
+                                        boolean visible = Boolean.valueOf(visibilityOverride);
+                                        top.setVisibilityOverride(v.getName(), visible);
+                                    }
+                                });
 
-                			if (importee.isImportDataPolicies()) {
-                				for (DataPolicy dp : child.getDataPolicies()) {
-                					DataPolicyMetadata role = (DataPolicyMetadata)dp;
-                					if (top.addDataPolicy(role) != null) {
-										throw new MojoExecutionException(top.getName() + "." + top.getVersion()
-												+ " imports a conflicting role " + role.getName() + " from "
-												+ child.getName() + "." + child.getVersion());
-                					}
-                					if (role.isGrantAll()) {
-                						role.setSchemas(child.getModelMetaDatas().keySet());
-                					}
-                				}
-                			}
-                            matched = importee;
-                            break;
+                                child.getOverrideTranslatorsMap().forEach((k, v) -> top.addOverideTranslator(v));
+
+                                if (importee.isImportDataPolicies()) {
+                                    for (DataPolicy dp : child.getDataPolicies()) {
+                                        DataPolicyMetadata role = (DataPolicyMetadata) dp;
+                                        if (top.addDataPolicy(role) != null) {
+                                            throw new MojoExecutionException(top.getName() + "." + top.getVersion()
+                                                    + " imports a conflicting role " + role.getName() + " from "
+                                                    + child.getName() + "." + child.getVersion());
+                                        }
+                                        if (role.isGrantAll()) {
+                                            role.setSchemas(child.getModelMetaDatas().keySet());
+                                        }
+                                    }
+                                }
+                                matched = importee;
+                                break;
+                            }
+                        }
+                        if (matched != null) {
+                            top.getVDBImports().remove(matched);
                         }
                     }
-                    if (matched != null) {
-                        top.getVDBImports().remove(matched);
-                    }
                 }
+                add(archive, "", directories.toArray(new File[directories.size()]));
+
+                File finalVDB = new File(this.outputDirectory.getPath(), "vdb.xml");
+                finalVDB.getParentFile().mkdirs();
+                VDBMetadataParser.marshell(top, new FileOutputStream(finalVDB));
+                addFile(archive, "META-INF/vdb.xml", finalVDB);
+
+            } catch (Exception e) {
+                throw new MojoExecutionException("Exception when creating artifact archive.", e);
             }
-            add(archive, "", directories.toArray(new File[directories.size()]));
-
-            File finalVDB = new File("target", "vdb.xml");
-            finalVDB.getParentFile().mkdirs();
-            VDBMetadataParser.marshell(top, new FileOutputStream(finalVDB));
-            addFile(archive, "META-INF/vdb.xml", finalVDB);
-
-        } catch (Exception e) {
-            throw new MojoExecutionException("Exception when creating artifact archive.", e);
+        } catch (MojoExecutionException e) {
+            throw new MojoExecutionException("Error running the vdb-maven-plugin.", e);
+        } finally {
+            Thread.currentThread().setContextClassLoader(oldCL);
         }
     }
 
@@ -174,7 +189,7 @@ public class VdbMojo extends AbstractMojo {
         if (f.exists() && f.isDirectory()) {
             File[] list = f.listFiles();
 
-            for (File l : list) {
+            for (File l : Objects.requireNonNull(list)) {
                 if (l.isDirectory()) {
                     directories.add(l);
                 }
@@ -185,17 +200,17 @@ public class VdbMojo extends AbstractMojo {
         }
     }
 
-    private File unzipContents(Artifact d) throws FileNotFoundException, IOException {
-        File f = new File("target", d.getArtifactId());
+    private File unzipContents(Artifact d) throws IOException {
+        File f = new File(this.outputDirectory.getPath(), d.getArtifactId());
         f.mkdirs();
-        System.out.println("unzipping " + d.getArtifactId() + " to directory "+ f.getCanonicalPath());
+        getLog().info("unzipping " + d.getArtifactId() + " to directory " + f.getCanonicalPath());
 
         byte[] buffer = new byte[1024];
         ZipInputStream zis = new ZipInputStream(new FileInputStream(d.getFile()));
         ZipEntry ze = zis.getNextEntry();
         while (ze != null) {
             String fileName = ze.getName();
-            System.out.println("\t" + fileName);
+            getLog().info("\t" + fileName);
             File newFile = new File(f, fileName);
             new File(newFile.getParent()).mkdirs();
             FileOutputStream fos = new FileOutputStream(newFile);
@@ -213,21 +228,19 @@ public class VdbMojo extends AbstractMojo {
     }
 
     private File getVDBFile() {
-        File f= new File(this.vdbXmlFile);
-        if (f.exists()) {
-            System.out.println("Found VDB = " + this.vdbXmlFile);
-            return f;
+        if (this.vdbXmlFile.exists()) {
+            getLog().info("Found VDB = " + this.vdbXmlFile);
+            return this.vdbXmlFile;
         } else {
-            f = new File(this.vdbFolder);
-            if (f.exists() && f.isDirectory()) {
-                File[] list = f.listFiles(new FilenameFilter() {
+            if (this.vdbFolder.exists() && this.vdbFolder.isDirectory()) {
+                File[] list = this.vdbFolder.listFiles(new FilenameFilter() {
                     @Override
                     public boolean accept(File dir, String name) {
                         return name.endsWith("-vdb.xml") || name.endsWith("-vdb.ddl");
                     }
                 });
-                if (list.length != 0) {
-                    System.out.println("Found VDB = " + list[0].getName());
+                if ((list != null ? list.length : 0) != 0) {
+                    getLog().info("Found VDB = " + list[0].getName());
                     return list[0];
                 }
             }
@@ -236,7 +249,7 @@ public class VdbMojo extends AbstractMojo {
     }
 
     private void addFile(ArchiveOutputStream archive, String name, File file) throws IOException {
-        System.out.println("Adding file = " + name +" from " + file.getCanonicalPath());
+        getLog().info("Adding file = " + name + " from " + file.getCanonicalPath());
         ArchiveEntry entry = this.entry(file, name);
         archive.putArchiveEntry(entry);
         IOUtils.copy(new FileInputStream(file), archive);
@@ -250,7 +263,7 @@ public class VdbMojo extends AbstractMojo {
             }
             String name = path + file.getName();
             if (file.isDirectory()) {
-                this.add(archive, name + SLASH, file.listFiles());
+                this.add(archive, name + SLASH, Objects.requireNonNull(file.listFiles()));
             } else {
                 if (!name.endsWith("vdb.xml")) {
                     addFile(archive, name, file);
@@ -259,10 +272,27 @@ public class VdbMojo extends AbstractMojo {
         }
     }
 
+    @SuppressWarnings("unchecked")
+    private ClassLoader getClassLoader() throws MojoExecutionException {
+        try {
+            List<URL> pathUrls = new ArrayList<>();
+            for (String mavenCompilePath : project.getCompileClasspathElements()) {
+                pathUrls.add(new File(mavenCompilePath).toURI().toURL());
+            }
+
+            URL[] urlsForClassLoader = pathUrls.toArray(new URL[pathUrls.size()]);
+            getLog().debug("urls for URLClassLoader: " + Arrays.asList(urlsForClassLoader));
+
+            // need to define parent classloader which knows all dependencies of the plugin
+            return new URLClassLoader(urlsForClassLoader, VdbMojo.class.getClassLoader());
+        } catch (Exception e) {
+            throw new MojoExecutionException("Couldn't create a classloader.", e);
+        }
+    }
+
     protected ArchiveOutputStream getStream(File artifact) throws IOException {
-        File outdir = new File(outputDirectory);
-        if (!outdir.exists()) {
-            outdir.mkdirs();
+        if (!this.outputDirectory.exists()) {
+            this.outputDirectory.mkdirs();
         }
         FileOutputStream output = new FileOutputStream(artifact);
         try {
@@ -273,7 +303,6 @@ public class VdbMojo extends AbstractMojo {
     }
 
     protected ArchiveEntry entry(File file, String name) {
-        ZipArchiveEntry entry = new ZipArchiveEntry(file, name);
-        return entry;
+        return new ZipArchiveEntry(file, name);
     }
 }

